row_test.py

```diff
@@ -103,3 +102,59 @@ def test_get_job_files_returns_the_page_size_for_the_second_page(storage_client)
     assert len(jobs) == 10
     assert jobs[0] == "page2"
     page_one_blob.assert_not_called()
+
+
+def test_get_circles_finds_circles():
+    image = root / "five_circles_with_text.png"
+
+    circles = row.get_circles(image, None)
+
+    assert circles is not None
+
+    assert len(circles) == 5
+
+
+def test_get_circles_ignores_non_image_files():
+    image = root / "five_circles_with_text.txt"
+
+    circles = row.get_circles(image, None)
+
+    assert circles is not None
+
+    assert len(circles) == 0
+
+
+def test_get_circles_saves_images_to_output():
+    image = root / "five_circles_with_text.png"
+    output = Path(__file__).parent / "data" / "nested"
+
+    row.get_circles(image, output)
+
+    assert output.exists()
+
+    assert len(list(output.glob("*.jpg"))) == 5
+
+
+def test_clean_ocr_text_removes_newlines():
+    text = """this is a
+
+test"""
+
+    assert row.clean_ocr_text(text) == "thisisatest"
+
+    text = """this is a               test"""
+    assert row.clean_ocr_text(text) == "thisisatest"
+
+
+def test_get_characters_finds_characters():
+    for item_path in root.glob("crop_*"):
+        file_name = item_path.name
+
+        image_array = np.frombuffer(item_path.read_bytes(), dtype=np.uint8)
+
+        characters = row.get_characters(image_array)
+
+        file_name = file_name.replace("crop_", "").replace("_", ":").replace(".jpg", "")
+
+        assert characters is not None
+        assert characters == file_name
```

Comment:
```suggestion
        expected_characters = file_name.replace("crop_", "").replace("_", ":").replace(".jpg", "")

        assert characters is not None
        assert characters == expected_characters
```
It's a little confusing to reuse the `file_name` variable, I recommend changing this to `expected_characters`